File: src/eQuantic.UI.Components.Shared/Avatar.cs
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Components.Shared;

/// <summary>Presence tiers (spec B6): Online = Success fill, Offline = TextMuted.</summary>
public enum PresenceStatus : byte
{
    None = 0,
    Online = 1,
    Offline = 2,
}

/// <summary>
/// The design system's Avatar (spec B6): sizes 24/32/40/56 (S/M/L/XL), Radius.Full — never square
/// for people. v1 renders the INITIALS fallback tier: a Subtle variant pair tinted deterministically
/// from the name (the spec's 2-stop gradient hash upgrades this when gradient tokens land); the
/// image tier joins with the Image primitive, the status dot with Stack.
/// </summary>
public sealed class Avatar : StatelessComponent
{
    private static readonly Variant[] TintPalette =
    [
        Variant.Primary, Variant.Success, Variant.Info, Variant.Warning, Variant.Destructive,
    ];

    public Avatar(string initials, SizeVariant size = SizeVariant.Medium, string? name = null)
    {
        Initials = initials;
        Size = size;
        Name = name;
    }

    /// <summary>Up to 2 characters (clipped per spec).</summary>
    public string Initials { get; init; }

    public SizeVariant Size { get; init; }

    /// <summary>Tint source (deterministic hash) — falls back to the initials when omitted.</summary>
    public string? Name { get; init; }

    /// <summary>Presence dot (spec B6): size/3.3, Success/TextMuted fill, 2dp Surface ring,
    /// bottom-end anchor — realized through Stack + Positioned.</summary>
    public PresenceStatus Status { get; init; }

    public override VisualNode Build(ComponentContext context)
    {
        var theme = context.Theme;
        var side = Size switch
        {
            SizeVariant.Small => 24f,
            SizeVariant.Medium => 32f,
            SizeVariant.Large => 40f,
            _ => 56f,
        };
        // Initials scale with the tier: 10/13/16/22 at 600 (derived — the spec pins only the sides).
        var labelSize = Size switch
        {
            SizeVariant.Small => 10f,
            SizeVariant.Medium => 13f,
            SizeVariant.Large => 16f,
            _ => 22f,
        };

        var seed = Name ?? Initials;
        var tint = theme.Colors(TintPalette[seed.Length % TintPalette.Length]);
        var clipped = Initials.Length > 2 ? Initials.Substring(0, 2) : Initials;

        var label = new Text(clipped, TypeRole.Caption, tint.OnSubtle, maxLines: 1)
        {
            StyleOverride = new TypeStyle(labelSize, labelSize, FontWeight.SemiBold, 0, 1.3f),
        };
        var content = new Row(gap: 0) { Main = MainAlign.Center, Height = SizeValue.Fill };
        content.Add(label);

        var circle = new Box(new BoxStyle
        {
            Width = side,
            Height = side,
            Background = tint.Subtle,
            CornerRadius = new CornerRadii(Radius.Full),
        }, content);

        if (Status == PresenceStatus.None) return circle;

        var dotSide = MathF.Round(side / 3.3f);
        var dotFill = Status == PresenceStatus.Online ? theme.Colors(Variant.Success).Base : theme.TextMuted;
        var dot = new Box(new BoxStyle
        {
            Width = dotSide,
            Height = dotSide,
            Background = dotFill,
            CornerRadius = new CornerRadii(Radius.Full),
            BorderWidth = 2f,
            BorderColor = theme.Surface,
        });

        var stack = new Stack();
        stack.Add(circle);
        stack.Add(new Positioned(dot, bottom: 0, end: 0));
        return stack;
    }
}

File: src/eQuantic.UI.Runtime/src/shared/components/Avatar.ts
import { $eq, Box, BoxStyle, BuildContext, CornerRadii, Icon, Image, Positioned, SizeVariantValue, Sizing, Stack, StatelessComponent, Text, TypeStyle, VariantValue, VisualNode } from "../runtime-exports";

export class Avatar extends StatelessComponent {
    static tintPalette: VariantValue[] = ['primary', 'success', 'info', 'warning', 'destructive'];
    declare initials: string;
    declare imageSource: any;
    declare size: SizeVariantValue;
    declare name: any;
    declare status: string;

    constructor(initials?: any, size: any = 'medium', name: any = null, props?: any) {
        super();
        if (initials !== undefined) this.initials = initials;
        if (size !== undefined) this.size = size;
        if (name !== undefined) this.name = name;
        if (this.size === undefined) this.size = 'small';
        if (this.status === undefined) this.status = 'none';
        this.initials = initials;
        this.size = size;
        this.name = name;
        if (props && typeof props === 'object') Object.assign(this, props);
    }

    build(context: BuildContext) {
        let theme = context.theme;
        let side = Sizing.avatar(this.size);
        let labelSize = (() => { const _s = this.size; if (_s === 'small') return 10; if (_s === 'medium') return 13; if (_s === 'large') return 16; return 22; })();
        let source: any; 
        if ((source = this.imageSource) != null) {
            let photo = new Image(source, side, side, 'cover', this.name ?? this.initials, { cornerRadius: new CornerRadii(theme.shape('full')) });
            return this.status === 'none' ? photo : this.withStatusDot(photo, side, theme);
        }
        let seed = this.name ?? this.initials;
        let tint = theme.colors(Avatar.tintPalette[seed.length % Avatar.tintPalette.length]);
        let clipped = this.initials.length > 2 ? $eq.text.substring(this.initials, 0, 2) : this.initials;
        let hasInitials = clipped.length > 0;
        let glyphSize = (() => { const _s = this.size; if (_s === 'small') return 16; if (_s === 'medium') return 20; if (_s === 'large') return 24; return 32; })();
        let face = hasInitials ? new Text(clipped, 'caption', tint.onSubtle, 1, 'start', false, false, null, 0, { styleOverride: new TypeStyle(labelSize, labelSize, 'semiBold', 0, Math.fround(1.3)) }) : new Icon('person', glyphSize, theme.textMuted);
        let content = face.centered();
        let circle = new Box(new BoxStyle({ width: side, height: side, background: hasInitials ? tint.subtle : theme.surfaceSubtle, cornerRadius: new CornerRadii(theme.shape('full')) }), content);
        return this.status === 'none' ? circle : this.withStatusDot(circle, side, theme);
    }

    withStatusDot(face: VisualNode, side: number, theme: any) {
        let dotSide = $eq.math.round(side / Math.fround(3.3));
        let dotFill = this.status === 'online' ? theme.colors('success').base : theme.textMuted;
        let dot = new Box(new BoxStyle({ width: dotSide, height: dotSide, background: dotFill, cornerRadius: new CornerRadii(theme.shape('full')), borderWidth: 2, borderColor: theme.surface }));
        let stack = new Stack();
        stack.add(face);
        stack.add(new Positioned(dot, null, 0, 0));
        return stack;
    }
}

